benchmarks/distributed/rpc/parameter_server/servers/AverageParameterServer.py

```diff
@@ -0,0 +1,96 @@
+import threading
+
+import torch
+import torch.distributed.rpc as rpc
+
+from .AverageParameterServerBase import AverageParameterServerBase
+from .ParameterServerBase import ParameterServerBase
+
+
+class AverageParameterServer(AverageParameterServerBase):
+
+    def __init__(
+        self,
+        rank,
+        trainer_count,
+        use_cuda_rpc
+    ):
+        super().__init__(rank)
+
+        self.lock = threading.Lock()
+
+        self.rank = rank
+        self.trainer_count = trainer_count
+        self.use_cuda_rpc = use_cuda_rpc
+
+        self.batch_number = 0
+        self.futures = {}
+        self.gradient_dict = {}
+
+    @staticmethod
+    def reset_state(ps_rref):
+        self = ps_rref.local_value()
+        self.batch_number = 0
+        self.futures.clear()
+        self.gradient_dict.clear()
+        self.clear_metrics()
+
+    @staticmethod
+    def get_metrics_rpc(ps_rref):
+        self = ps_rref.local_value()
+        return self.get_metrics()
+
+    def param_key(self, param_loc):
+        return f"{self.batch_number},{param_loc}"
+
+    def clear_batch_state(self):
+        self.futures.clear()
+        self.gradient_dict.clear()
+
+    def process_gradient(self, gradient, param_loc):
+        if param_loc not in self.gradient_dict:
+            self.record_straggler_start(self.param_key(param_loc))
+            self.record_batch_start(self.param_key(param_loc))
+            self.gradient_dict[param_loc] = gradient
+        else:
+            self.gradient_dict[param_loc] += gradient
+
+    @ParameterServerBase.record_method(name="average computation")
+    def average(self, param_loc):
+        param_loc_avg = self.gradient_dict[param_loc]
+        param_loc_avg / (1.0 * self.trainer_count)
+        return param_loc_avg
+
+    @staticmethod
+    @rpc.functions.async_execution
+    def average_gradient(
+        ps_rref,
+        received_batch_number,
+        param_loc,
+        gradient
+    ):
+        self = ps_rref.local_value()
+        if type(gradient) is list:
+            gradient = self.sparse_rpc_format_to_tensor(gradient)
+        if not self.use_cuda_rpc:
```

Comment:
if the purpose of this `if` clause is to avoid unintentional cpu->cuda copy, we can drop the branch and always call `gradient.cuda(self.rank)`, as it will be a no-op if it is already on the target cuda device.

Reply:
okay.